src/runtime_src/xocl/core/memory.cpp

```diff
@@ -287,6 +287,41 @@ update_memidx_nolock(const device* device, const buffer_object_handle& boh)
 }
 
 // private
+memory::memidx_type
+memory::
+get_grpidx_nolock(const device* dev) const
+{
+  int cuidx = -1;
+  int argidx = -1;
+
+  // already initialized
+  if (m_memidx>=0) 
+    return m_memidx;
+
+  if (m_karg.empty())
+    return -1;
+
+  for (auto& karg : m_karg) {
+    auto kernel = karg.first;
+    auto arg = karg.second;
+    if (kernel) {
+      argidx = arg;
+      cuidx = kernel->get_uid();
+      break;
+    }
+  }
+
+  if ((cuidx == -1) || (argidx == -1))
+    return -1;
+
+  auto device = xrt_core::get_userpf_device(dev->get_handle());
+  m_memidx = device->get_groupIndex(cuidx, argidx);
+  if (m_memidx < 0)
+    return -1;
+
+  return m_memidx;
+}
+
```

Comment:
Let's not change if we can avoid

Reply:
Btw.  Errors must be exceptions, do not return -1 unless that is a valid non error return value.

Reply:
I have updated this function.  Please review and share your your feedback.